File: app/src/main/java/com/app/ralaunch/controls/editor/manager/ControlEditDialogVisibilityManager.java
package com.app.ralaunch.controls.editor.manager;

import android.view.View;

import androidx.annotation.NonNull;

import com.app.ralaunch.R;
import com.app.ralaunch.controls.ControlData;

/**
 * 控件编辑对话框选项可见性管理器
 * 统一管理根据控件类型和形状动态显示/隐藏选项的逻辑
 */
public class ControlEditDialogVisibilityManager {
    
    /**
     * 选项可见性规则接口
     */
    public interface VisibilityRule {
        /**
         * 判断选项是否应该显示
         * @param data 控件数据
         * @return true 表示显示，false 表示隐藏
         */
        boolean shouldShow(ControlData data);
    }
    
    /**
     * 更新外观选项的可见性
     * @param view 外观视图
     * @param data 控件数据
     */
    public static void updateAppearanceOptionsVisibility(@NonNull View view, @NonNull ControlData data) {
        
        // 圆角半径：仅矩形形状的按钮和文本控件显示（摇杆是圆形，不需要圆角）
        View cardCornerRadius = view.findViewById(R.id.card_corner_radius);
        if (cardCornerRadius != null) {
            boolean isRectangle = (data.shape == ControlData.SHAPE_RECTANGLE);
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            // 摇杆类型不显示圆角半径（摇杆始终是圆形）
            cardCornerRadius.setVisibility((isRectangle && !isJoystick) ? View.VISIBLE : View.GONE);
        }
        
        // 摇杆透明度：仅摇杆类型显示
        View cardStickOpacity = view.findViewById(R.id.card_stick_opacity);
        if (cardStickOpacity != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            cardStickOpacity.setVisibility(isJoystick ? View.VISIBLE : View.GONE);
        }
        
        // 摇杆圆心大小：仅摇杆类型显示
        View cardStickKnobSize = view.findViewById(R.id.card_stick_knob_size);
        if (cardStickKnobSize != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            cardStickKnobSize.setVisibility(isJoystick ? View.VISIBLE : View.GONE);
        }
        
    }
    
    /**
     * 更新键值设置选项的可见性
     * @param view 键值设置视图
     * @param data 控件数据
     */
    public static void updateKeymapOptionsVisibility(@NonNull View view, @NonNull ControlData data) {
        // 普通按钮的键值设置：仅按钮类型显示（文本控件和摇杆不支持按键映射）
        View itemKeyMapping = view.findViewById(R.id.item_key_mapping);
        View itemToggleMode = view.findViewById(R.id.item_toggle_mode);
        if (itemKeyMapping != null || itemToggleMode != null) {
            boolean isButton = (data.type == ControlData.TYPE_BUTTON);
            // 文本控件和摇杆不支持按键映射，隐藏键值设置
            boolean isText = (data.type == ControlData.TYPE_TEXT);
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            // 只有按钮类型（非文本、非摇杆）才显示键值设置
            boolean shouldShow = isButton && !isText && !isJoystick;
            if (itemKeyMapping != null) {
                itemKeyMapping.setVisibility(shouldShow ? View.VISIBLE : View.GONE);
            }
            if (itemToggleMode != null) {
                itemToggleMode.setVisibility(shouldShow ? View.VISIBLE : View.GONE);
            }
        }
    }
    
    /**
     * 更新基本信息选项的可见性
     * @param view 基本信息视图
     * @param data 控件数据
     */
    public static void updateBasicInfoOptionsVisibility(@NonNull View view, @NonNull ControlData data) {
        // 控件名称：文本控件隐藏（避免与文本内容重复）
        View itemControlName = view.findViewById(R.id.item_control_name);
        if (itemControlName != null) {
            boolean isText = (data.type == ControlData.TYPE_TEXT);
            itemControlName.setVisibility(isText ? View.GONE : View.VISIBLE);
        }
        
        // 文本内容：仅文本控件显示
        View itemTextContent = view.findViewById(R.id.item_text_content);
        if (itemTextContent != null) {
            boolean isText = (data.type == ControlData.TYPE_TEXT);
            itemTextContent.setVisibility(isText ? View.VISIBLE : View.GONE);
        }
        
        // 摇杆模式选择：仅摇杆类型显示
        View itemJoystickMode = view.findViewById(R.id.item_joystick_mode);
        if (itemJoystickMode != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            itemJoystickMode.setVisibility(isJoystick ? View.VISIBLE : View.GONE);
        }
        
        // 摇杆左右选择：仅摇杆类型且为SDL控制器模式或鼠标模式时显示
        // - 鼠标模式：左摇杆=相对移动，右摇杆=八方向攻击（通过组合键的鼠标左键）
        // - SDL控制器模式：左摇杆=左摇杆输入，右摇杆=右摇杆输入
        View itemJoystickStickSelect = view.findViewById(R.id.item_joystick_stick_select);
        if (itemJoystickStickSelect != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            boolean isSDLControllerMode = (data.joystickMode == ControlData.JOYSTICK_MODE_SDL_CONTROLLER);
            boolean isMouseMode = (data.joystickMode == ControlData.JOYSTICK_MODE_MOUSE);
            itemJoystickStickSelect.setVisibility((isJoystick && (isSDLControllerMode || isMouseMode)) ? View.VISIBLE : View.GONE);
        }
        
        // 右摇杆攻击模式：仅摇杆类型且为鼠标模式且为右摇杆时显示
        View itemRightStickAttackMode = view.findViewById(R.id.item_right_stick_attack_mode);
        if (itemRightStickAttackMode != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            boolean isMouseMode = (data.joystickMode == ControlData.JOYSTICK_MODE_MOUSE);
            boolean isRightStick = data.xboxUseRightStick;
            itemRightStickAttackMode.setVisibility((isJoystick && isMouseMode && isRightStick) ? View.VISIBLE : View.GONE);
        }
        
        // 鼠标移动范围：仅摇杆类型且为鼠标模式且为右摇杆时显示
        View itemMouseRange = view.findViewById(R.id.item_mouse_range);
        if (itemMouseRange != null) {
            boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
            boolean isMouseMode = (data.joystickMode == ControlData.JOYSTICK_MODE_MOUSE);
            boolean isRightStick = data.xboxUseRightStick;
            itemMouseRange.setVisibility((isJoystick && isMouseMode && isRightStick) ? View.VISIBLE : View.GONE);
        }
    }
    
    /**
     * 更新所有选项的可见性
     * @param basicInfoView 基本信息视图
     * @param appearanceView 外观视图
     * @param keymapView 键值设置视图
     * @param data 控件数据
     */
    public static void updateAllOptionsVisibility(@NonNull View basicInfoView,
                                                  @NonNull View appearanceView, 
                                                  @NonNull View keymapView,
                                                  @NonNull ControlData data) {
        updateBasicInfoOptionsVisibility(basicInfoView, data);
        updateAppearanceOptionsVisibility(appearanceView, data);
        updateKeymapOptionsVisibility(keymapView, data);
    }
    
    /**
     * 更新所有选项的可见性（兼容旧方法，不包含基本信息视图）
     * @param appearanceView 外观视图
     * @param keymapView 键值设置视图
     * @param data 控件数据
     */
    public static void updateAllOptionsVisibility(@NonNull View appearanceView, 
                                                  @NonNull View keymapView,
                                                  @NonNull ControlData data) {
        updateAppearanceOptionsVisibility(appearanceView, data);
        updateKeymapOptionsVisibility(keymapView, data);
    }
    
    /**
     * 根据规则设置视图可见性
     * @param view 视图
     * @param data 控件数据
     * @param rule 可见性规则
     */
    public static void setVisibilityByRule(@NonNull View view, 
                                          @NonNull ControlData data,
                                          @NonNull VisibilityRule rule) {
        if (view != null) {
            view.setVisibility(rule.shouldShow(data) ? View.VISIBLE : View.GONE);
        }
    }
    
    /**
     * 创建基于控件类型的可见性规则
     */
    public static VisibilityRule createTypeRule(int targetType) {
        return data -> data.type == targetType;
    }
    
    /**
     * 创建基于控件形状的可见性规则
     */
    public static VisibilityRule createShapeRule(int targetShape) {
        return data -> data.shape == targetShape;
    }
    
    /**
     * 创建组合规则（AND）
     */
    public static VisibilityRule and(VisibilityRule... rules) {
        return data -> {
            for (VisibilityRule rule : rules) {
                if (!rule.shouldShow(data)) {
                    return false;
                }
            }
            return true;
        };
    }
    
    /**
     * 创建组合规则（OR）
     */
    public static VisibilityRule or(VisibilityRule... rules) {
        return data -> {
            for (VisibilityRule rule : rules) {
                if (rule.shouldShow(data)) {
                    return true;
                }
            }
            return false;
        };
    }
    
    /**
     * 创建取反规则（NOT）
     */
    public static VisibilityRule not(VisibilityRule rule) {
        return data -> !rule.shouldShow(data);
    }
}

File: app/src/main/java/com/app/ralaunch/controls/editors/managers/ControlEditDialogVisibilityManager.kt
package com.app.ralaunch.controls.editors.managers

import android.view.View
import com.app.ralaunch.controls.data.ControlData

/**
 * 控件编辑对话框可见性管理器
 * 使用统一的规则容器管理各个选项的可见性逻辑
 */
object ControlEditDialogVisibilityManager {

    //region Helper Classes

    /**
     * 可见性规则
     * @param isVisible 判断是否可见的lambda函数
     * @param resourceIds 需要应用该规则的资源ID列表
     */
    data class Rule(
        val isVisible: (ControlData) -> Boolean,
        val resourceIds: List<Int>
    )

    /**
     * 规则容器
     * 统一管理所有可见性规则
     */
    class RuleContainer {
        private val rules = mutableListOf<Rule>()

        /**
         * 添加规则
         */
        fun addRule(rule: Rule): RuleContainer {
            rules.add(rule)
            return this
        }

        /**
         * 添加规则（便捷方法）
         */
        fun addRule(
            isVisible: (ControlData) -> Boolean,
            vararg resourceIds: Int
        ): RuleContainer {
            return addRule(Rule(isVisible, resourceIds.toList()))
        }

        /**
         * 应用所有规则到视图
         */
        fun applyRules(view: View, data: ControlData) {
            rules.forEach { rule ->
                val visible = rule.isVisible(data)
                val visibility = if (visible) View.VISIBLE else View.GONE

                rule.resourceIds.forEach { resourceId ->
                    view.findViewById<View?>(resourceId)?.visibility = visibility
                }
            }
        }

        /**
         * 清空所有规则
         */
        fun clear() {
            rules.clear()
        }
    }

    //endregion

    /**
     * 规则容器实例
     */
    private val ruleContainer = RuleContainer().apply {
        // 圆角半径（仅在矩形形状时显示）
        addRule(
            {
                when (it) {
                    is ControlData.Button -> it.shape == ControlData.Button.Shape.RECTANGLE
                    is ControlData.Text -> it.shape == ControlData.Text.Shape.RECTANGLE
                    is ControlData.TouchPad -> true
                    is ControlData.MouseWheel -> true
                    else -> false
                }
            },
            com.app.ralaunch.R.id.card_corner_radius
        )

        // 摇杆中心透明度（仅摇杆类型显示）
        addRule(
            { it is ControlData.Joystick },
            com.app.ralaunch.R.id.card_stick_opacity
        )

        // 摇杆圆心大小（仅摇杆类型显示）
        addRule(
            { it is ControlData.Joystick },
            com.app.ralaunch.R.id.card_stick_knob_size
        )

        // 形状选择（仅按钮和文本控件显示）
        addRule(
            { it is ControlData.Button || it is ControlData.Text },
            com.app.ralaunch.R.id.item_control_shape
        )

        // 摇杆模式选择（仅摇杆类型显示）
        addRule(
            { it is ControlData.Joystick },
            com.app.ralaunch.R.id.item_joystick_mode
        )

        // 文本内容编辑（仅文本控件显示）
        addRule(
            { it is ControlData.Text },
            com.app.ralaunch.R.id.item_text_content
        )

        // 普通键值设置（仅按钮显示，文本控件不支持键值映射）
        addRule(
            { it is ControlData.Button },
            com.app.ralaunch.R.id.item_key_mapping
        )

        // 摇杆键值设置（仅摇杆显示，且为键盘模式）
        addRule(
            {
                it is ControlData.Joystick && it.mode == ControlData.Joystick.Mode.KEYBOARD
            },
            com.app.ralaunch.R.id.item_joystick_key_mapping
        )

        // 切换模式（仅按钮显示）
        addRule(
            { it is ControlData.Button },
            com.app.ralaunch.R.id.item_toggle_mode
        )

        // 文本透明度（按钮和文本控件显示）
        addRule(
            { it is ControlData.Button || it is ControlData.Text },
            com.app.ralaunch.R.id.card_text_opacity
        )

        // 右摇杆攻击模式（仅摇杆鼠标模式显示）
        addRule(
            {
                it is ControlData.Joystick && it.mode == ControlData.Joystick.Mode.MOUSE
            },
            com.app.ralaunch.R.id.item_right_stick_attack_mode
        )

        // 鼠标范围和速度（摇杆鼠标模式或触摸板显示，不包括鼠标滚轮）
        addRule(
            {
                (it is ControlData.Joystick && it.mode == ControlData.Joystick.Mode.MOUSE) ||
                it is ControlData.TouchPad
            },
            com.app.ralaunch.R.id.item_mouse_range
        )

        // 双击模拟摇杆（仅触摸板显示）
        addRule(
            { it is ControlData.TouchPad },
            com.app.ralaunch.R.id.layout_double_click_joystick
        )

        // 鼠标滚轮方向（仅鼠标滚轮显示）
        addRule(
            { it is ControlData.MouseWheel },
            com.app.ralaunch.R.id.card_mousewheel_orientation
        )

        // 鼠标滚轮反转（仅鼠标滚轮显示）
        addRule(
            { it is ControlData.MouseWheel },
            com.app.ralaunch.R.id.card_mousewheel_reverse
        )

        // 鼠标滚轮灵敏度（仅鼠标滚轮显示）
        addRule(
            { it is ControlData.MouseWheel },
            com.app.ralaunch.R.id.card_mousewheel_sensitivity
        )

        // 鼠标滚轮速度倍率（仅鼠标滚轮显示）
        addRule(
            { it is ControlData.MouseWheel },
            com.app.ralaunch.R.id.card_mousewheel_ratio
        )

        // 摇杆左右选择（仅在手柄模式或鼠标模式时显示）
        addRule(
            {
                it is ControlData.Joystick && (
                    it.mode == ControlData.Joystick.Mode.GAMEPAD ||
                    it.mode == ControlData.Joystick.Mode.MOUSE
                )
            },
            com.app.ralaunch.R.id.item_joystick_stick_select
        )
    }

    /**
     * 更新所有选项的可见性
     */
    fun updateAllOptionsVisibility(view: View, data: ControlData) {
        ruleContainer.applyRules(view, data)
    }
}

